Save session before the `location` header is sent to the browser

patriksimek opened the pull request:
This PR fixes the problem described in https://github.com/expressjs/session/pull/69. When user calls `res.redirect()`, it checks whether the session needs to be saved (or touched) before the `location` header is sent do the browser.

This fix doesn't work for situation when user sets the `location` header manually.

Test included.

dougwilson commented:
Please propose a solution that is not specific to Express.js; it should work when users set the `Location` header manually; in fact, this is broken because it'll add `res.redirect()` even when someone is not using Express and if they call it, it'll fail.

dougwilson commented on index.js:
```diff
@@ -199,6 +199,70 @@ function session(options){
+        return;
+      }
+      
+      return _redirect.call(res, url);
```
This throws is the user is not using Express; this library works outside on Express in things like Connect as well.

dougwilson commented on index.js:
```diff
@@ -199,6 +199,70 @@ function session(options){
+    // proxy redirect() to commit the session
+    var _redirect = res.redirect;
+    var redirected = false;
+    res.redirect = function redirect(url) {
```
We should not be adding `res.redirect` if it doesn't already exist.

dougwilson replied:
This will break everyone in Express doing `res.redirect(301, url)`, which is a valid call.

patriksimek commented:
Thanks for the review, now it's not specific to Express and also it should correctly pass all the arguments to proxied method.

I'm not sure if it is possible to fix this for situation when user sets the `location` header manually. It can't be done in `setHeader` neither in `end` method. I'm out of ideas here.

dougwilson commented:
> and also it should correctly pass all the arguments to proxied method.

Can you add tests for this, then? To me, that does not seem to be the case at all. The following all need to be tested to work here: `res.redirect(308, url)`, `res.redirect(url, 308)`, and `res.redirect(url)`. The tests should also validate that the status code is 308.

dougwilson commented:
> Thanks for the review, now it's not specific to Express

Well, if it's not specific to Express, how will people without `res.redirect` not still encounter the issue?

> I'm not sure if it is possible to fix this for situation when user sets the location header manually. It can't be done in setHeader neither in end method. I'm out of ideas here.

Right, it's really hard, which is why the issue is still outstanding. If someone can come up with a solution it'll be accepted :) The goal of the modules in the `expressjs` organization is not to cater only to Express, but rather work in any framework that accepts Express-like middleware. I'm also not too keen on the large code duplication (which needs a lot more tests, as indicated from the dropped code coverage :( ).

patriksimek commented:
So, as you mentioned [here](https://github.com/expressjs/session/pull/69#issuecomment-57092528), the only solution is to buffer the response. I have updated my PR with this kind of patch that should work as neutral middleware. I have also added an option for this feature so it could be published as non-breaking feature.

Code duplication and test coverage should be fine now.

dougwilson commented on index.js:
```diff
@@ -199,10 +203,61 @@ function session(options){
+          debug('redirect found, buffering response')
+          buffered = true;
+          bufferedChunks = new Buffer(0);
+          writeHeadLater = function() { _writeHead.call(res, statusCode, reason); };
```
The contract with `.writeHead` is that we must call it sync right away; we cannot actually delay it's call, sadly. This is a limitation with injecting ourselves directly into the http response object without an abstraction.

The reason is what once code calls `res.writeHead()`, then Node.js guarantees that the side-effects have occurred sync, i.e the following must work, but this PR breaks it:
```js
res.writeHead(500)
assert(res.statusCode === 500)
```
```

patriksimek commented:
I have updated the code with fix for `statusCode`, but if the problem is in delaying `writeHead`, than you can close this right now :)

dougwilson commented:
The status code was just an example of the issue. For instance, that may "fix" the status code, but the following will still fail:
```js
res.writeHead(500)
assert(res.headersSent === 500)
```

as well as:
```js
res.writeHead(204)
assert(res.chunkedEncoding === false)
```

which can induce very subtle bugs in existing applications.

dougwilson commented on index.js:
```diff
@@ -199,10 +203,63 @@ function session(options){
+        headWritten = true;
 
+        // search for location header only for 3xx status codes
+        if (statusCode < 300 || statusCode >= 400) {
```
Won't this not work if the user calls the following, since you are not looking at the already-set response code?
```js
res.statusCode = 301
res.setHeader('Location', '/')
res.end()
```

patriksimek replied:
It should be fine, implicit `writeHead` is always called with already-set response code.

https://github.com/nodejs/io.js/blob/3777f415625ce538de5edbb19f7330356da190a8/lib/_http_server.js#L150

This case is also covered by added test `should buffer the response #2`.

patriksimek commented:
I agree this feature is a breaking one so it should not be enabled by default. Users could be notified via docs about fact that enabling this feature may cause some properties to not being set immediately. Just an idea...

dougwilson commented on index.js:
```diff
@@ -199,10 +203,63 @@ function session(options){
+      res.write = _write = function write(chunk, encoding, callback) {
+        if (!headWritten) res.writeHead(this.statusCode);
+        
+        if (buffered) {
```
So this code is assuming that calling `res.writeHead` will invoke your `res.writeHead` sync, but we are purposely violating this, so in the spirit of that, we need to assume that someone else replaced `res.writeHead` that is also not sync (for example, what if they are using two `express-session` modules?).

dougwilson commented on index.js:
```diff
@@ -199,10 +203,63 @@ function session(options){
+        if (buffered) {
+          bufferedChunks = Buffer.concat([bufferedChunks, !Buffer.isBuffer(chunk) ? new Buffer(chunk, encoding) : chunk]);
+          if ('function' === typeof(callback)) setImmediate(callback);
+          return true;
```
Return should be `false`, which means "hey, stop writing to me", which is the case here, as we're bufferring.

dougwilson commented on index.js:
```diff
@@ -199,10 +203,63 @@ function session(options){
+        
+        if (buffered) {
+          bufferedChunks = Buffer.concat([bufferedChunks, !Buffer.isBuffer(chunk) ? new Buffer(chunk, encoding) : chunk]);
+          if ('function' === typeof(callback)) setImmediate(callback);
```
The `callback` should only be called once we actually performed the write, so these need to be buffered up and called once we write out our buffer.

dougwilson commented on index.js:
```diff
@@ -199,10 +203,63 @@ function session(options){
+        if (!headWritten) res.writeHead(this.statusCode);
+        
+        if (buffered) {
+          bufferedChunks = Buffer.concat([bufferedChunks, !Buffer.isBuffer(chunk) ? new Buffer(chunk, encoding) : chunk]);
```
Over the long haul, `Buffer.concat` can get extremely slow. Why not just make `bufferedChunks` an array and store them up?

andrukas commented:
Hello,

does these fixes will be released?

dougwilson commented:
I'm going to close this since nothing has been changed from the issues pointed out in comments over a year ago.
